src/registrar/models/user.py

```diff
@@ -230,68 +200,51 @@ def get_formatted_name(self):
     def has_contact_info(self):
         return bool(self.title or self.email or self.phone)
 
-    def clean(self):
-        """Extends clean method to perform additional validation, which can raise errors in django admin."""
-        super().clean()
-
-        if self.portfolio is None and self._get_portfolio_permissions():
-            raise ValidationError("When portfolio roles or additional permissions are assigned, portfolio is required.")
-
-        if self.portfolio is not None and not self._get_portfolio_permissions():
-            raise ValidationError("When portfolio is assigned, portfolio roles or additional permissions are required.")
-
-    def _get_portfolio_permissions(self):
-        """
-        Retrieve the permissions for the user's portfolio roles.
-        """
-        portfolio_permissions = set()  # Use a set to avoid duplicate permissions
-
-        if self.portfolio_roles:
-            for role in self.portfolio_roles:
-                if role in self.PORTFOLIO_ROLE_PERMISSIONS:
-                    portfolio_permissions.update(self.PORTFOLIO_ROLE_PERMISSIONS[role])
-        if self.portfolio_additional_permissions:
-            portfolio_permissions.update(self.portfolio_additional_permissions)
-        return list(portfolio_permissions)  # Convert back to list if necessary
-
-    def _has_portfolio_permission(self, portfolio_permission):
+    def _has_portfolio_permission(self, portfolio, portfolio_permission):
         """The views should only call this function when testing for perms and not rely on roles."""
 
-        if not self.portfolio:
+        if not portfolio:
             return False
 
-        portfolio_permissions = self._get_portfolio_permissions()
+        portfolio_perms = self.portfolio_permissions.filter(portfolio=portfolio, user=self).first()
+        if not portfolio_perms:
+            return False
 
+        portfolio_permissions = portfolio_perms._get_portfolio_permissions()
         return portfolio_permission in portfolio_permissions
```

Comment:
```suggestion
        user_portfolio_perms = self.portfolio_permissions.filter(portfolio=portfolio, user=self).first()
        if not user_portfolio_perms:
            return False

        return portfolio_permission in user_portfolio_perms._get_portfolio_permissions()
```
Nit-picky suggestion. I found these lines confusing since the names were so similar between portfolio_perms and portfolio_permissions.